native/dgebal.go

```diff
@@ -0,0 +1,227 @@
+// If job is lapack.Permute, only permuting will be done.
+// If job is lapack.Scale, only scaling will be done.
+// If job is lapack.PermuteScale, both permuting and scaling will be done.
+//
+// On return, if job is lapack.Permute or lapack.PermuteScale, it will hold that
+//  A[i,j] == 0,   for i > j and j∈{0, ..., ilo-1, ihi+1, ..., n-1}.
+// If job is lapack.None or lapack.Scale, or if n == 0, it will hold that
+//  ilo == 0 and ihi == n-1.
+//
+// On return, scale will contain information about the permutations and scaling
+// factors applied to A. If π(j) denotes the index of the column interchanged
+// with column j, and D[j,j] denotes the scaling factor applied to column j,
+// then
+//  scale[j] == π(j),     for j∈{0, ..., ilo-1, ihi+1, ..., n-1},
+//           == D[j,j],   for j∈{ilo, ..., ihi}.
+//
+// Dgebal is an internal routine. It is exported for testing purposes.
+func (impl Implementation) Dgebal(job lapack.Job, n int, a []float64, lda int, scale []float64) (ilo, ihi int) {
+	switch job {
+	default:
+		panic(badJob)
+	case lapack.None, lapack.Permute, lapack.Scale, lapack.PermuteScale:
+	}
+	checkMatrix(n, n, a, lda)
+	if len(scale) != n {
+		panic("lapack: bad length of scale")
+	}
+
+	ilo = 0
+	ihi = n - 1
+
+	if n == 0 || job == lapack.None {
+		for i := range scale {
+			scale[i] = 1
+		}
+		return ilo, ihi
+	}
+
+	bi := blas64.Implementation()
+	swapped := true
+
+	if job == lapack.Scale {
+		goto Scaling
+	}
+
+	// Permutation to isolate eigenvalues if possible.
+	//
+	// Search for rows isolating an eigenvalue and push them down.
+	for swapped {
+		swapped = false
+	RowLoop:
+		for i := ihi; i >= 0; i-- {
+			for j := 0; j <= ihi; j++ {
+				if i == j {
+					continue
+				}
+				if a[i*lda+j] != 0 {
+					continue RowLoop
+				}
+			}
+			// Row i has only zero off-diagonal elements in the
+			// block A[ilo:ihi+1,ilo:ihi+1].
+			scale[ihi] = float64(i)
+			if i != ihi {
+				bi.Dswap(ihi+1, a[i:], lda, a[ihi:], lda)
+				bi.Dswap(n, a[i*lda:], 1, a[ihi*lda:], 1)
+			}
+			ihi--
+			if ihi == 0 {
+				scale[0] = 1
+				return ilo, ihi
+			}
+			swapped = true
+			break
+		}
+	}
+	// Search for columns isolating an eigenvalue and push them left.
+	swapped = true
+	for swapped {
+		swapped = false
+	ColLoop:
+		for j := ilo; j <= ihi; j++ {
+			for i := ilo; i <= ihi; i++ {
+				if i == j {
+					continue
+				}
+				if a[i*lda+j] != 0 {
+					continue ColLoop
+				}
+			}
+			// Column j has only zero off-diagonal elements in the
+			// block A[ilo:ihi+1,ilo:ihi+1].
+			scale[ilo] = float64(j)
+			if j != ilo {
+				bi.Dswap(ihi+1, a[j:], lda, a[ilo:], lda)
+				bi.Dswap(n-ilo, a[j*lda+ilo:], 1, a[ilo*lda+ilo:], 1)
+			}
+			swapped = true
+			ilo++
+			break
+		}
+	}
+
+Scaling:
+	for i := ilo; i <= ihi; i++ {
+		scale[i] = 1
+	}
+
+	if job == lapack.Permute {
+		return ilo, ihi
+	}
+
+	// Balance the submatrix in rows ilo to ihi.
+
```

Comment:
I like to separate comments by a blank line if they relate to a larger piece of code below, not only to the subsequent block. I'll keep it here.